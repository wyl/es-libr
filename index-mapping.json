[
  {
    "index": "caas-cn-zaobao-online",
    "mapper": {
      "data.context.displaySetting.displayAssetsMultiple.keyword": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_displaySetting_displayAssetsMultiple_keyword"
      },
      "data.context.displaySetting.zbsgFlag": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_displaySetting_zbsgFlag"
      },
      "data.context.tags.name.keyword": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_tags_name_keyword"
      },
      "data.context.tags.urlPath.keyword": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_tags_urlPath_keyword"
      },
      "data.context.tags.urlPath": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_tags_urlPath"
      },
      "data.context.tags.type.keyword": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_tags_type_keyword"
      },
      "data.context.tags.type": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_tags_type"
      },
      "data.resolution.section.uniqueName.keyword": {
        "type": {
          "type": "text"
        },
        "reName": "data_resolution_section_uniqueName_keyword"
      },
      "data.context.updated": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_updated"
      },
      "data.context.id": {
        "type": {
          "type": "keyword"
        },
        "reName": "data_context_id"
      },
      "data.context.title": {
        "type": {
          "type": "text",
          "analyzer": "ik_smart"
        },
        "reName": "data_context_title"
      },
      "data.context.type": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_type"
      },
      "data.context.elements.children": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_elements_children"
      },
      "data.context.authors.profiles.content.id": {
        "type": {
          "type": "text"
        },
        "reName": "data_context_authors_profiles_content_id"
      },
      "data.context.tags.id": {
        "type": {
          "type": "keyword"
        },
        "reName": "data_context_tags_id"
      }
    }
  }
]
