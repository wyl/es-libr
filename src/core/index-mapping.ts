import fs from 'fs'

export type IndexMappings = Record<string, IndexMappingTool>

type IndexMapping = {
  index: string
  mapper: FieldsMapper
}
type FieldsMapper = {
  [key: string]: {
    type: Record<string, unknown>
    reName: string
  }
}

async function loadIndexMapping(filePath: string) {
  let indexMappings: IndexMappings = {}
  try {
    const tmp = JSON.parse(
      fs.readFileSync(filePath, 'utf-8'),
    ) as Array<IndexMapping>
    indexMappings = Object.fromEntries(
      tmp.map((it) => {
        return [it.index, new IndexMappingTool(it)]
      }),
    )
  } catch (error) {
    throw new Error(`load index mapping error from "${filePath}" : ` + error)
  }
  return () => indexMappings
}

class IndexMappingTool {
  protected index: string
  protected mapper: FieldsMapper

  constructor(mapping: IndexMapping) {
    this.index = mapping.index
    this.mapper = mapping.mapper
  }

  buildCreateIndexBody() {
    return this.mapper
  }

  mapping() {
    return Object.fromEntries(
      Object.entries(this.mapper).map(([key, value]) => {
        return [key, value.reName]
      }),
    )
  }
}

export { loadIndexMapping }
