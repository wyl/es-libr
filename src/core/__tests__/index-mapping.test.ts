import { loadIndexMapping } from "../index-mapping";

const failedPaths = ["error path", "undefined", ""];
const successPaths = ["index-mapping.json"];

describe("loadIndexMapping Testing", () => {
  test.each(failedPaths)("Wrong path [%s]", (path) => {
    expect(loadIndexMapping(path)).rejects.toThrow(
      "load index mapping error from"
    );
  });

  test.each(successPaths)("Success [%s] ", async (path) => {
    const mapping = (await loadIndexMapping(path))();

    expect(mapping).toBeTruthy();
    expect(mapping).toBeInstanceOf(Object);
    expect(mapping).toHaveProperty("caas-cn-zaobao-online");
    expect(mapping).toHaveProperty(
      "caas-cn-zaobao-online.index",
      "caas-cn-zaobao-online"
    );
    expect(mapping).toHaveProperty("caas-cn-zaobao-online.mapping");
    expect(mapping).toMatchSnapshot();

    // Object.entries(mapping["caas-cn-zaobao-online"].mapping()).forEach(
    //   ([key, value]) => {
    //     expect(value).not.toBe(value);
    //   }
    // );
  });
});
