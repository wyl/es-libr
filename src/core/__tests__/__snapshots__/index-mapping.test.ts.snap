// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIndexMapping Testing Success [index-mapping.json]  1`] = `
{
  "caas-cn-zaobao-online": IndexMappingTool {
    "index": "caas-cn-zaobao-online",
    "mapper": {
      "data.context.authors.profiles.content.id": {
        "reName": "data_context_authors_profiles_content_id",
        "type": {
          "type": "text",
        },
      },
      "data.context.displaySetting.displayAssetsMultiple.keyword": {
        "reName": "data_context_displaySetting_displayAssetsMultiple_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.displaySetting.zbsgFlag": {
        "reName": "data_context_displaySetting_zbsgFlag",
        "type": {
          "type": "text",
        },
      },
      "data.context.elements.children": {
        "reName": "data_context_elements_children",
        "type": {
          "type": "text",
        },
      },
      "data.context.id": {
        "reName": "data_context_id",
        "type": {
          "type": "keyword",
        },
      },
      "data.context.tags.id": {
        "reName": "data_context_tags_id",
        "type": {
          "type": "keyword",
        },
      },
      "data.context.tags.name.keyword": {
        "reName": "data_context_tags_name_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.tags.type.keyword": {
        "reName": "data_context_tags_type_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.tags.urlPath.keyword": {
        "reName": "data_context_tags_urlPath_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.title": {
        "reName": "data_context_title",
        "type": {
          "analyzer": "ik_smart",
          "type": "text",
        },
      },
      "data.context.type": {
        "reName": "data_context_type",
        "type": {
          "type": "text",
        },
      },
      "data.context.updated": {
        "reName": "data_context_updated",
        "type": {
          "type": "text",
        },
      },
      "data.resolution.section.uniqueName.keyword": {
        "reName": "data_resolution_section_uniqueName_keyword",
        "type": {
          "type": "text",
        },
      },
    },
  },
}
`;

exports[`loadIndexMapping Testing Success Testing Success [index-mapping.json]  1`] = `
{
  "caas-cn-zaobao-online": IndexMappingTool {
    "index": "caas-cn-zaobao-online",
    "mapper": {
      "data.context.authors.profiles.content.id": {
        "reName": "data_context_authors_profiles_content_id",
        "type": {
          "type": "text",
        },
      },
      "data.context.displaySetting.displayAssetsMultiple.keyword": {
        "reName": "data_context_displaySetting_displayAssetsMultiple_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.displaySetting.zbsgFlag": {
        "reName": "data_context_displaySetting_zbsgFlag",
        "type": {
          "type": "text",
        },
      },
      "data.context.elements.children": {
        "reName": "data_context_elements_children",
        "type": {
          "type": "text",
        },
      },
      "data.context.id": {
        "reName": "data_context_id",
        "type": {
          "type": "keyword",
        },
      },
      "data.context.tags.id": {
        "reName": "data_context_tags_id",
        "type": {
          "type": "keyword",
        },
      },
      "data.context.tags.name.keyword": {
        "reName": "data_context_tags_name_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.tags.type.keyword": {
        "reName": "data_context_tags_type_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.tags.urlPath.keyword": {
        "reName": "data_context_tags_urlPath_keyword",
        "type": {
          "type": "text",
        },
      },
      "data.context.title": {
        "reName": "data_context_title",
        "type": {
          "analyzer": "ik_smart",
          "type": "text",
        },
      },
      "data.context.type": {
        "reName": "data_context_type",
        "type": {
          "type": "text",
        },
      },
      "data.context.updated": {
        "reName": "data_context_updated",
        "type": {
          "type": "text",
        },
      },
      "data.resolution.section.uniqueName.keyword": {
        "reName": "data_resolution_section_uniqueName_keyword",
        "type": {
          "type": "text",
        },
      },
    },
  },
}
`;
